Gm diff manager

mgalbier opened the pull request:
- Correct getFileContent to avoid matching of fileName with directory names
- Add handling for working directory
- Use test command to call DiffManager
- Add slower zooming for changed components
- Change behaviour of CDiff to use internally the complete sha1 hashes if available
  -> This is to avoid ambiguity conflicts with git intern objects other than commits
  <a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23discussion_r61048637%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23discussion_r61048863%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23discussion_r61049714%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23discussion_r61054536%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23discussion_r61054617%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23discussion_r61054686%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23discussion_r61056860%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23discussion_r61057052%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23issuecomment-214689914%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23issuecomment-214709662%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23issuecomment-214710103%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23issuecomment-214689914%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Alright%2C%20I%27ve%20seen%20it%20all%2C%20just%20fix%20the%20few%20small%20issues%20and%20I%27ll%20merge%20it.%5Cr%5Cn%5Cr%5CnThanks.%22%2C%20%22created_at%22%3A%20%222016-04-26T10%3A00%3A20Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Updated%22%2C%20%22created_at%22%3A%20%222016-04-26T11%3A24%3A47Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/17177211%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mgalbier%22%7D%7D%2C%20%7B%22body%22%3A%20%22Thanks.%22%2C%20%22created_at%22%3A%20%222016-04-26T11%3A27%3A12Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20643ff8b0e8eab0c1da7775bb0f409bb310850988%20FilePersistence/src/version_control/GitPiecewiseLoader.cpp%2011%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23discussion_r61048863%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Looks%20fine%20to%20me.%20Does%20it%20work%20as%20expected%3F%22%2C%20%22created_at%22%3A%20%222016-04-26T08%3A47%3A30Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22As%20far%20as%20I%20can%20tell%3A%20yes.%22%2C%20%22created_at%22%3A%20%222016-04-26T09%3A27%3A52Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/17177211%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mgalbier%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20FilePersistence/src/version_control/GitPiecewiseLoader.cpp%3AL41-58%22%7D%2C%20%22Pull%208dbf7f6bcff02fd1da02b30621b392ac968f75b7%20VersionControlUI/src/DiffManager.h%2015%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23discussion_r61049714%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22A%20slightly%20better%20name%20is%20%60findAllItemsWithParentIn%60%5Cr%5Cn%5Cr%5CnAre%20you%20talking%20about%20direct%20parents%20or%20Ancestors%3F%20If%20the%20second%20then%20please%20use%20the%20appropriate%20word.%22%2C%20%22created_at%22%3A%20%222016-04-26T08%3A54%3A08Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ah%20you%20are%20right%2C%20ancestor%20would%20be%20correct.%5Cr%5Cn%5Cr%5CnI%27ll%20change%20it.%22%2C%20%22created_at%22%3A%20%222016-04-26T09%3A28%3A21Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/17177211%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mgalbier%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/DiffManager.h%3AL92-103%22%7D%2C%20%22Pull%204d84618f37dad36513bb6da20cf23c9e5d6521a7%20FilePersistence/src/version_control/Commit.cpp%205%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23discussion_r61048637%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22This%20check%20is%20still%20somewhat%20brittle%2C%20as%20it%20still%20relies%20on%20specifics%20of%20the%20name%2C%20which%20might%20change%20in%20the%20future.%5Cr%5Cn%5Cr%5CnWhy%20not%20use%20%60QFileInfo%3A%3AisDir%60%20for%20entries%20that%20contain%20%60fileName%60%3F%20This%20should%20be%20more%20stable.%22%2C%20%22created_at%22%3A%20%222016-04-26T08%3A45%3A53Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Hm%2C%20I%20already%20avoid%20matching%20directories%20by%20using%20%60QFileInfo%3A%3AfileName%28%29%60%20which%20returns%20just%20the%20filename%20without%20the%20path.%5Cr%5Cn%5Cr%5CnThe%20check%20afterwards%20is%20to%20see%20if%20it%20is%20an%20UID%20which%20can%20be%20matched%20using%20contains%20or%20something%20else%2C%20where%20contains%20would%20not%20be%20good%20%28file%20%5C%22foobar%5C%22%20contains%20%5C%22foo%5C%22%29%20and%20thus%20I%20use%20exact%20matching%20for%20just%20the%20filename.%5Cr%5Cn%5Cr%5CnBut%20I%20agree%20it%20is%20brittle.%22%2C%20%22created_at%22%3A%20%222016-04-26T09%3A27%3A24Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/17177211%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/mgalbier%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20FilePersistence/src/version_control/Commit.cpp%3AL82-95%22%7D%2C%20%22Pull%208dbf7f6bcff02fd1da02b30621b392ac968f75b7%20VersionControlUI/src/DiffManager.cpp%2049%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23discussion_r61057052%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Know%20your%20libraries%2C%20instead%20of%20these%20several%20lines%2C%20use%20simply%3A%5Cr%5Cnhttp%3A//doc.qt.io/qt-5/qset.html%23subtract%22%2C%20%22created_at%22%3A%20%222016-04-26T09%3A43%3A37Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/DiffManager.cpp%3AL261-298%22%7D%2C%20%22Pull%208dbf7f6bcff02fd1da02b30621b392ac968f75b7%20VersionControlUI/src/DiffManager.cpp%206%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/371%23discussion_r61056860%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22The%20name%20of%20this%20local%20field%20is%20out%20of%20place.%20Nothing%20in%20this%20method%20itself%20suggests%20what%20the%20resulting%20list%20will%20be%20used.%20You%27re%20just%20computing%20which%20are%20the%20items%20that%20have%20parents%2C%20and%20somebody%20on%20the%20outside%20uses%20this%20to%20remove%20some%20items%20from%20somewhere.%5Cr%5Cn%5Cr%5CnCall%20%20this%20%60result%60%22%2C%20%22created_at%22%3A%20%222016-04-26T09%3A42%3A22Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20VersionControlUI/src/DiffManager.cpp%3AL174-200%22%7D%7D%7D'></a>
  <a href='https://www.codereviewhub.com/'><img src='http://www.codereviewhub.com/site/github-bar.png' height=40></a>
- [ ] <a href='#crh-comment-Pull 4d84618f37dad36513bb6da20cf23c9e5d6521a7 FilePersistence/src/version_control/Commit.cpp 5'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/371#discussion_r61048637'>File: FilePersistence/src/version_control/Commit.cpp:L82-95</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> This check is still somewhat brittle, as it still relies on specifics of the name, which might change in the future.
  Why not use `QFileInfo::isDir` for entries that contain `fileName`? This should be more stable.
- <a href='https://github.com/mgalbier'><img border=0 src='https://avatars.githubusercontent.com/u/17177211?v=3' height=16 width=16'></a> Hm, I already avoid matching directories by using `QFileInfo::fileName()` which returns just the filename without the path.
  The check afterwards is to see if it is an UID which can be matched using contains or something else, where contains would not be good (file "foobar" contains "foo") and thus I use exact matching for just the filename.
  But I agree it is brittle.
- [ ] <a href='#crh-comment-Pull 643ff8b0e8eab0c1da7775bb0f409bb310850988 FilePersistence/src/version_control/GitPiecewiseLoader.cpp 11'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/371#discussion_r61048863'>File: FilePersistence/src/version_control/GitPiecewiseLoader.cpp:L41-58</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Looks fine to me. Does it work as expected?
- <a href='https://github.com/mgalbier'><img border=0 src='https://avatars.githubusercontent.com/u/17177211?v=3' height=16 width=16'></a> As far as I can tell: yes.
- [ ] <a href='#crh-comment-Pull 8dbf7f6bcff02fd1da02b30621b392ac968f75b7 VersionControlUI/src/DiffManager.h 15'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/371#discussion_r61049714'>File: VersionControlUI/src/DiffManager.h:L92-103</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> A slightly better name is `findAllItemsWithParentIn`
  Are you talking about direct parents or Ancestors? If the second then please use the appropriate word.
- <a href='https://github.com/mgalbier'><img border=0 src='https://avatars.githubusercontent.com/u/17177211?v=3' height=16 width=16'></a> Ah you are right, ancestor would be correct.
  I'll change it.
- [ ] <a href='#crh-comment-Pull 8dbf7f6bcff02fd1da02b30621b392ac968f75b7 VersionControlUI/src/DiffManager.cpp 6'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/371#discussion_r61056860'>File: VersionControlUI/src/DiffManager.cpp:L174-200</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> The name of this local field is out of place. Nothing in this method itself suggests what the resulting list will be used. You're just computing which are the items that have parents, and somebody on the outside uses this to remove some items from somewhere.
  Call  this `result`
- [ ] <a href='#crh-comment-Pull 8dbf7f6bcff02fd1da02b30621b392ac968f75b7 VersionControlUI/src/DiffManager.cpp 49'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/371#discussion_r61057052'>File: VersionControlUI/src/DiffManager.cpp:L261-298</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Know your libraries, instead of these several lines, use simply:
  http://doc.qt.io/qt-5/qset.html#subtract
- [ ] <a href='#crh-comment-General Comment'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/371#issuecomment-214689914'>General Comment</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Alright, I've seen it all, just fix the few small issues and I'll merge it.
  Thanks.
- <a href='https://github.com/mgalbier'><img border=0 src='https://avatars.githubusercontent.com/u/17177211?v=3' height=16 width=16'></a> Updated
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Thanks.

<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/371?mark_as_completed=1'><img src='http://www.codereviewhub.com/site/github-mark-as-completed.png' height=26></a>&nbsp;<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/371?approve=1'><img src='http://www.codereviewhub.com/site/github-approve.png' height=26></a>&nbsp;<a href='https://github.com/dimitar-asenov/Envision/pull/371'><img src='http://www.codereviewhub.com/site/github-refresh.png' height=26></a>
<a href='#crh-end'></a>

mgalbier committed:
Correct getFileContent to avoid matching of fileName with directory names

mgalbier committed:
Add handling for working directory

mgalbier committed:
Use test command to call DiffManager

mgalbier committed:
Add slower zooming for changed components

mgalbier committed:
Change behaviour of CDiff to use internally the complete sha1 hashes if available

This is to avoid ambiguity conflicts with git intern objects other than commits

dimitar-asenov commented:
This pull request is now managed with CodeReviewHub. <a href='https://github.com/dimitar-asenov/Envision/pull/371'>Refresh</a>.<a href='#crh-update-none'></a>

dimitar-asenov commented on FilePersistence/src/version_control/Commit.cpp:
```diff
@@ -82,8 +82,13 @@ bool Commit::getFileContent(QString fileName, const char*& content, int& content
 	{
 		iter = files_.constBegin();
-		while (!iter.key().contains(fileName) && iter != files_.constEnd())
+		while (iter != files_.constEnd())
```
This check is still somewhat brittle, as it still relies on specifics of the name, which might change in the future.

Why not use `QFileInfo::isDir` for entries that contain `fileName`? This should be more stable.

mgalbier replied:
Hm, I already avoid matching directories by using `QFileInfo::fileName()` which returns just the filename without the path.

The check afterwards is to see if it is an UID which can be matched using contains or something else, where contains would not be good (file "foobar" contains "foo") and thus I use exact matching for just the filename.

But I agree it is brittle.

dimitar-asenov commented on FilePersistence/src/version_control/GitPiecewiseLoader.cpp:
```diff
@@ -41,8 +41,17 @@ GitPiecewiseLoader::~GitPiecewiseLoader() {}
+
+	bool isWorkDir = (revision_ == GitRepository::WORKDIR);
+	// in case of workdir just use regular grep
+	// TODO is grep called correctly like this?
```
Looks fine to me. Does it work as expected?

mgalbier replied:
As far as I can tell: yes.

dimitar-asenov commented on VersionControlUI/src/DiffManager.h:
```diff
@@ -91,6 +92,11 @@ class VERSIONCONTROLUI_API DiffManager
+		/**
+		 * Returns all items which have any of their parents present in \a items.
+		 */
+		static QSet<Visualization::Item*> findAllItemsWithParentPresent(QSet<Visualization::Item*> items);
```
A slightly better name is `findAllItemsWithParentIn`

Are you talking about direct parents or Ancestors? If the second then please use the appropriate word.

mgalbier replied:
Ah you are right, ancestor would be correct.

I'll change it.

dimitar-asenov commented on VersionControlUI/src/DiffManager.cpp:
```diff
@@ -174,6 +174,26 @@ QSet<Model::Node*> DiffManager::findAllNodesWithDirectParentPresent(QHash<Model:
 
+QSet<Visualization::Item*> DiffManager::findAllItemsWithParentPresent(QSet<Visualization::Item*> items)
+{
+	QSet<Visualization::Item*> itemsMarkedForRemoval;
```
The name of this local field is out of place. Nothing in this method itself suggests what the resulting list will be used. You're just computing which are the items that have parents, and somebody on the outside uses this to remove some items from somewhere.

Call  this `result`

dimitar-asenov commented on VersionControlUI/src/DiffManager.cpp:
```diff
@@ -239,11 +261,37 @@ void DiffManager::createOverlaysForChanges(Visualization::ViewItem* diffViewItem
+
+	QSet<Visualization::Item*> removeItems = findAllItemsWithParentPresent(allItemsToScale);
+
+	auto it = allItemsToScale.begin();
```
Know your libraries, instead of these several lines, use simply:
http://doc.qt.io/qt-5/qset.html#subtract

dimitar-asenov commented:
Alright, I've seen it all, just fix the few small issues and I'll merge it.

Thanks.

mgalbier committed:
Add isDir check

mgalbier committed:
Rename findAllItemsWithParentPresent to findAllItemsWithAncestorIn
Rename local variable itemsMarkedForRemoval to result
Use subtract method instead of manually deleting entries

mgalbier commented:
Updated

dimitar-asenov commented:
Thanks.
